nupic/research/frameworks/vernon/experiments/self_supervised_experiment.py

```diff
@@ -0,0 +1,498 @@
+from torch.backends import cudnn
+from torch.optim.lr_scheduler import OneCycleLR
+from torch.utils.data import DataLoader, RandomSampler
+
+from nupic.research.frameworks.pytorch.lr_scheduler import ComposedLRScheduler
+from nupic.research.frameworks.pytorch.self_supervised_utils import EncoderClassifier
+from nupic.research.frameworks.vernon.experiment_utils import create_lr_scheduler
+from nupic.research.frameworks.vernon.experiments import SupervisedExperiment
+from nupic.research.frameworks.vernon.network_utils import create_model
+
+try:
+    from apex import amp
+except ImportError:
+    amp = None
+
+
+__all__ = ["SelfSupervisedExperiment"]
+
+
+# Improves performance when using fixed size images (224) and CNN
+cudnn.benchmark = True
+
+
+class SelfSupervisedExperiment(SupervisedExperiment):
+    """
+    General experiment class used to train neural networks in self-supervised learning
+    tasks.
+
+    Self-supervised experiments have three important dataset splits: unsupervised,
+    supervised, and validation. The unsupervised set consists of unlabeled data for
+    representation learning, the supervised set consists of a typically smaller
+    amount of labeled data for which to train a classifier, and the test set is used
+    to evaluate the classifier.
+
+    The model passed into this experiment under the name 'model' should be some kind
+    of encoder which extends torch.nn.Module and implements both forward() and encode():
+
+    class YourEncoderHere(torch.nn.Module):
+            # this forward pass is used during unsupervised training
+            def forward(self, x):
+                x = ...
+                return x
+
+            # the encode function is used during supervised training
+            def encode(self, x):
+                encoded = ...
+                return encoded
+
+    The reason for this requirement is that the loss for the unsupervised training
+    can (and usually does) operate on a different output than the representation
+    that is used during supervised training. For example, an autoencoder's forward()
+    should reconstruct its input, while its encode() function should yield the
+    intermediate representation of a typically lower dimension.
+
+    The model passed into this experiment under the name 'classifier' should take
+    the encoding provided by .encode() as input and output a classification in its
+    forward() method.
+
+    The validation step trains a new classifier on top of the existing frozen model,
+    and then proceeds to test this classifier on the test set. The number of
+    supervised training epochs to train for each validation is given by
+    supervised_training_epochs_per_validation.
+    """
+
+    def setup_experiment(self, config):
+        """
+        Configure the experiment for training.
+        :param config: Dictionary containing the configuration parameters
+
+            ********* Params from the parent class, SupervisedExperiment ***************
```

Comment:
You probably shouldn't copy all the config params from the parent class. It's likely to get out of date as soon as changes are made to the parent. Just mention the additions/changes in this class.

Reply:
changed, thanks